### ISSUE TITLE
Add binary input/output functions

### ISSUE THREAD

**Issue by roji** [ChristophBerg/postgresql-unit#21]
This extension seems to only support text input/output, and lacks binary input/output (receive/send) functions. Binary transfer would make transfer much more efficient on these types of values, and would make the extension usable in a wider range of contexts. For example, Npgsql (PostgreSQL driver for .NET) is a binary-first driver - support for text-only types is very limited (this is due mainly to PostgreSQL protocol limitations).

PS Great extension idea, thanks for doing this.

**Comment by zr40** [comment]
You say Npgsql requires binary format because of PostgreSQL protocol limitations? What are these limitations exactly? Most other drivers support text format just fine.

Note that binary format is not guaranteed to be portable across server versions. According to [the documentation](https://www.postgresql.org/docs/10/static/protocol-overview.html#PROTOCOL-FORMAT-CODES): "Keep in mind that binary representations for complex data types might change across server versions; the text format is usually the more portable choice."

**Comment by zr40** [comment]
By the way, if the user is currently unable to use unit because of this Npgsql binary-only restriction, they could work around the problem by casting to or from text. Or alternatively, they could use the `@@` operator to extract the number part and `*` to put it back in.

For example: `velocity_unit @@ 'm/s'` and `5 * 'm/s'::unit`

**Comment by roji** [comment]
> You say Npgsql requires binary format because of PostgreSQL protocol limitations? What are these limitations exactly? Most other drivers support text format just fine.

The primary difficulty with the PostgreSQL protocol is that each query has to specify, in advance, for each result column, whether it wants it back as text or binary. There is no way to for Npgsql to say, for example, "I always want `unit` to come back as text" - it has to be specified on each specific query sent to the database. Npgsql does expose a special API allowing the user to say "bring back this column as text", but this is a non-standard API feature (so hurts database portability), and users in general shouldn't be concerned with protocol details such as text vs. binary.

> As for extension types (such as postgres-unit), a driver cannot rely on a particular binary format if it is changed in a future version of the extension, since it has no reliable way of determining what the structure will be.

I don't really understand this one... Do you expect drivers to parse the textual representation for `unit` coming from your database? If so, why do you assume this textual representation will remain stable, while the binary one will not? It's your choice as the extension maintainer to estimate whether your wire representation will change, regardless of whether it's binary or text, isn't it?

Npgsql has been working for years with the binary representation, and PostgreSQL has never caused any breakage. If a change in the representation does occur, the driver can be modified to support the new format based on the server version, for example. In the case of the `jsonb` type, by the way, the binary representation contains a byte prefix that indicates the version - this is an explicit placeholder for supporting other binary wire representations, allowing the type to evolve and drivers with it.

> As for postgres-unit in particular, a hypothetical binary format would include the value (double precision) and an array of base dimensions (array of signed integers). The driver would have to reconstruct the displayed unit using these base dimensions, which essentially requires reimplementing that logic client-side.

That's exactly right, and that's what Npgsql does for the types it supports - including incredibly complex types (arrays of composites of anything). To me this is the proper and efficient way for a driver to interact with a database.

To better understand your view, how do you imagine drivers and users interacting with `unit` types? Does the driver simply hand the user a string such as `200 * 0.5 l`? What is the user supposed to do with that, assuming they want to access the data? If you expect them to parse it, isn't that something that a good driver should do on behalf of the user, returning a strongly-typed, highly-structured representation?

And to give a bit of history, a long time ago Npgsql actually did work with text - it would parse strings coming back from PostgreSQL, and construct them when sending parameters. This created many issues - text representations are sensitive to the session locale (the digit separator is different between English and French), it was extremely inefficient for many types (why parse a complex text representation of a timestamp when you can just read an efficient binary representation instead), etc. etc. Since we've switched to binary we've seen a big reduction in user issues and much better performance.

At the end of the day, the driver and the database are two programs talking to each other, there's no reason to have them parsing and constructing complex textual representations when they can talk in binary....

**Comment by roji** [comment]
> The binary format is nice for things that you want to transform into native types on the client side. For "unit", you'd first have to write an implementation, or else the binary representation is almost completely useless. (Assuming the binary protocol would just send the binary representation on the server side.) Also, the point of the "unit" type is that it uses nicely formatted strings (text!) for input/output, so forcing the user to use binary defeats that.

I completely agree. The only sense in providing a binary format is for drivers which provide a native type representation - this is exactly what an Npgsql plugin would do for your extension. You can think of it as an option for drivers which want to provide this "improved" level of support (which includes client-side types). It doesn't hurt anyone :)

>  Of course we could just send the text version over the binary protocol, but then you could really just retrieve the text format. (For any given query, you'll need to know the returned data types anyway, or else you can't use them in your program.)

I agree, this would be useless.

> Which representation would you want to see in the binary protocol?

Basically something very similar to your storage representation: a double precision value followed by followed by an array of exponents (8-byte signed). The exponents would be preceded by the number of exponents.

As you mentioned above, it's worth thinking backwards compatibility and any possible changes you're imagining for the future - it's true that the binary representation nails down more than the text representation does (e.g. the size of the exponents). Ideally the binary representation wouldn't change (although if it absolutely has to, that's not necessarily a big catastrophe as Npgsql and other drivers can find out which version of the extension if being used).

**Comment by dpsenner** [comment]
It's about time to join the conversation, being the origin of the issue. But first of all let me express great thanks to both of you for working on this issue.

A little bit of context always helps. We're evaluating methods to store measurements into the postgres database. Some of those values are going to be user input and storing such data as decimal + user input unit effectively defeats any rounding or converting issues. Storing values is a great thing but stored data is of no use if it is never used by anyone. So we are also going to read those values in the future. We are going to read the values as they are but we also see a usecase for reporting tools that do some calculations on the values. For now let it be sums and such, but it may eventually will become even more complicated when someone wants to do some statistical analysis on data.

When evaluating this we at first stumbled upon postgresql-unit and it surely reads as promising idea. We however also evaluate whether we should rather introduce one or more composite types in the database that could store the same as `unit` does but without requiring a plugin. I.e. something like the following provides about the same functionality as the `unit` datatype does:

```sql
CREATE TYPE unit2 AS (
    value double precision,
    unit_exponents char[]
);
```

@ChristophBerg I'm sure you've given this a lot of thoughts into this plugin an am intrigued to know why you chose to implement a plugin instead of defining composite types along with a set of operations. Would you please explain a few words how you came to the decision of implementing a plugin?

I ask because I do not work in the time wasting business. So if we find out that we are not going to set sails with postgresl-unit, there's no functional requirement to implement binary transfer.

Thanks and greets!

**Comment by dpsenner** [comment]
Surely those conversion features could be implemented as functions, possibly even functions that provide operators and operator overloads.

**Comment by roji** [comment]
@dpsenner uh, that's exactly what this extension is for, or am I misunderstanding? Who would reimplement these and where? On which type exactly?

**Comment by roji** [comment]
> @dpsenner: I'm thinking of adding send/recv functions that directly expose the internal binary format. The downside is that you can't use unit conversion/scaling for input/output. Would that help you?

From my point of view this is exactly what's needed. If I understand things correctly, you can still apply unit conversions expressed in SQL (e.g. `SELECT some_unit_column @ 'GB/d' AS traffic`) and then transfer the resulting unit in binary over the wire. This would solve the issue for Npgsql and would give the option of a more efficient value transfer for anyone that cares to use it.

**Comment by ChristophBerg** [comment]
The `@` operator yields "text" (it can't yield a unit because that's internally always scaled as "B/s"), but that would work, yes.

**Comment by ChristophBerg** [comment]
That's why I was reluctant to support binary - if you just "SELECT unitvalue" in the binary format, you'll lose all nice formatting, so you can't input '1000 m' and get back '1 km'. You'll have to rewrite all queries to do explicit reformatting. (And currently there's no function that autoscales the result with prefixes, you'll have to say `@ 'km'`, and you'll get '0.001 km' if it was only a meter.)

**Comment by ChristophBerg** [comment]
Re strong types, I'm considering adding typmods that will constrain the dimension to a fixed value, i.e. `unit(LENGTH)`. Would that help with your use case?

**Comment by ChristophBerg** [comment]
Fixed in 238afbde4841b26475916a3a316d2ebdbe204ed9. (With the wrong commit message, sorry...)
